### ISSUE TITLE
Prevent merging if one feature has tag present and other feature has tag null

### ISSUE THREAD

**Issue by slhh** [openstreetmap/iD#2935]
Merging of ways with different value of the same key is prohibited yet. Unfortunately, this was only the tip of the iceberg in view of corrupting the content of the database.
The semicolon separated value were not nice, but at least detectable errors; therefore, they were less critical than the remaining undectable errors.

A merge is safe if the meaning of the data is unchanged, or if the user is aware of the changes.
The requirement to protect againt conflicting values shows, that unexperienced users of iD don't check tags well when merging. Therefore, they aren't aware of any change to the meaning of the data being induced by the merging of tags. Either there must not be any change of the meaning of data, or iD needs to make the user aware of the change of the meaning.

The problem is: Using a tag, which is present in only one of the original ways is already a change of the meaning. The tag is likely invalid for the full length of the merged way.

There is an exception when the of the one way tag is an implied tag of the other way. In this case we can assume the merge is safe. 
I believe there is also an exception when merging with an untagged way which has been created in the same edit session. In this case spreading the tags is likely intentional, and we can assume the merge is safe.

**Comment by bhousel** [comment]
I really don't understand the issue. Could you dumb it down for me?

**Comment by slhh** [comment]
Assume we have two consecutive ways:

1.  Way: 9.975 km long, just a normal road with asphalt surface, being tagged as:
   * highway=tertiary
2. Way;  25m long, passing under an old bridge, being tagged as: 
   * highway=tertiary
   * surface=cobblestone
   * turn:forward=right
   * maxheight=3
   * smoothness=bad

Both ways are rendered equally in iD. Therefore an unexperienced user might merge the ways; maybe he wants to add a name for the full length.

The result of the merge is a 10 km long tertiary with bad cobblestone surface, height restrictions, right turn markings along its full length. This is wrong information, but all tags seems to be ok formally; therefore, the error is undetectable without local knowledge.

**Comment by tyrasd** [comment]
For comparison, this is the modal dialog JOSM shows in non-expert mode in such a case (in expert mode there is a tag conflict disambiguation dialog instead):

![combine confirmation_100](https://cloud.githubusercontent.com/assets/1927298/12612290/f0bf7942-c4f1-11e5-8745-a0943b2477bd.png)

**Comment by bhousel** [comment]
Ok, but we really need #2929 first, because this kind of merge is really common and useful.

**Comment by slhh** [comment]
@bhousel Which usecase do you have in mind for this kind of merge?
In my opinion the primary usecase is merging with a totally untagged line/area in order to extend the feature, but I had already suggested not to prevent merging in this case.

Another usecase for this kind of merge is a workaround for the missing copy tags operation, but we should not support this and implement #839 instead.

#2929 is not the right solution for any remaining usecases, because this would likely result in overusing the modifier key. The user might use the modifier key to intentionally override the tag protection, but accidentially overriding the relation protection too.

I see the following solutions for the remaining usecases:
* Let the user copy the tags first before merging (needs #839)
* Generate a warning massage showing which tags will be spread over the full length/area.
* Place the merge button somewhere, where a preview of the tagging effects can be shown before the merge operation is invoked by the user. #2924 would be one way to do this.

**Comment by bhousel** [comment]
> @bhousel Which usecase do you have in mind for this kind of merge?
> In my opinion the primary usecase is merging with a totally untagged line/area in order to extend the feature, but I had already suggested not to prevent merging in this case.

I use this merge often in TIGER cleanup to merge ways where one has a name and one does not.

It tends to happen in places like this.. The small segment was imported with no name, but it really should be merged with Poplar Creek Lane.

<img width="333" alt="screenshot 2016-03-11 10 12 47" src="https://cloud.githubusercontent.com/assets/38784/13706150/daf36c5c-e771-11e5-915c-8da23a1fa8a7.png">

**Comment by bhousel** [comment]
Closing for now.   I thought about this more and I perform merges like this all the time when editing.

A pretty common thing I do is to delete the conflicting tags and then merge.  For example merging a street with a misspelled name "1st Avenue" and "1st Ave" and absolutely no other conflicting tags - I'll just delete the name from "1st Ave", select them both, and hit <kbd>C</kbd> to merge.

**Comment by slhh** [comment]
#4334 is a similar issue with relations instead of tags.
@bhousel
>  Closing for now. I thought about this more and I perform merges like this all the time when editing.

This means keeping a bug, which is potentially destructive for OSM data, in order to use it as a workaround for a missing feature like #3028, which is exactly addressing your usecase. I might have described #3028 to general, which can result in an overwhelming display in some cases, but the availability feature can easily restricted to typical usecases of not more than two diffent geometry types. The display woun't be overwhelming in such usecases anyway, and the restriction enable some further simplfication of the display.
